### ISSUE TITLE
Training stuck at 0% after few epochs while training with DDP

### ISSUE THREAD

**Issue by HareshKarnan** [PyTorchLightning/pytorch-lightning#5865]
## 🐛 Bug

I recently updated to pytorch_lightning 1.1.7 and noticed that after a few epochs of training, the training % is stuck at 0% and never progresses. When I switched back to 1.1.4, this strange behavior does not occur. I do not know the root cause of this issue.

 - PyTorch Version (e.g., 1.0): 1.1.7
 - OS (e.g., Linux): Linux, U18
 - How you installed PyTorch (`conda`, `pip`, source): pip install 
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: RTX 2080 x3
 - Any other relevant information:

**Comment by ndrplz** [comment]
Related? I'm facing a similar issue, not sure if those might help
- https://github.com/PyTorchLightning/pytorch-lightning/issues/4612 
- https://github.com/pytorch/pytorch/issues/38702

**Comment by matyushinleonid** [comment]
Hi, @HareshKarnan @ndrplz. Folks, may I ask you to run your pipelines with `NCCL_ASYNC_ERROR_HANDLING=1`?
Seems like we have the same problem. I use `1.2.0.dev0` lightning version, so this issue may migrate to the new release... My error trace:

https://pastebin.com/LNxG2JF6

**Comment by HareshKarnan** [comment]
@ndrplz i dont know if it is related, but the problem here is that training does happen for first few epochs - in my case, it ran for 13 epochs and then gets stuck at 0% by epoch 14.

**Comment by matyushinleonid** [comment]
> 
> 
> https://pastebin.com/QhanSNrK
> 
> Here is my output. It gets stuck at 0% while training in epoch 16

run your script with `NCCL_ASYNC_ERROR_HANDLING=1` flag. Like `NCCL_ASYNC_ERROR_HANDLING=1 python train.py`. Without this flag, I do not get any error like you since, by default, this async error does not raise an Exception (at least, for me). If we have the same error traces, it will help community to fix the bug.

**Comment by HareshKarnan** [comment]
I ran the script again as ```NCCL_ASYNC_ERROR_HANDLING=1 python train.py``` and got no difference in the output : 

https://pastebin.com/jtLyuXTN

**Comment by genghisun** [comment]
same problem here, stuck at 0% at epoch 18

**Comment by Borda** [comment]
Mind share code ideally in Colab to reproduce?

**Comment by genghisun** [comment]
In my case, it stuck at 0% at epoch 18 with 2 gpus ddp before.
Then I try to use only 1 gpu, currently trained for 100+ epochs without any problem.

**Comment by stillwalker1234** [comment]
have the same issue after updating to 1.1.8, will try with 1.2.0.dev0 to see if it has the same error

**Comment by stillwalker1234** [comment]
1.2.0rc1 also has the issue, 1.1.6 does not

**Comment by SeanNaren** [comment]
Would you be able to try master? We've recently consolidated the branches back to master!

**Comment by tchaton** [comment]
Hey everyone,

Could it be related to this issue: https://github.com/PyTorchLightning/pytorch-lightning/pull/6004

Best,
T.C

**Comment by stillwalker1234** [comment]
Started a run with master + @tchaton 's patch, will see how it goes.

**Comment by Borda** [comment]
@tchaton any update here?

**Comment by talolard** [comment]
Also having this problem. 
Always getting stuck on epoch 2, if I have checkpoints enabled when training with ddp on single machine

**Comment by talolard** [comment]
Wanted to add some details. 
We ran some different code with copy paste and the problem reoccured. We had the checkpoint set up to monitor train_loss and once we took that out everything worked fine. 
In other words, I suspect this has something to do with monitor on ddp .

**Comment by edenlightning** [comment]
Thanks! Will take a look and try to resolve it soon.

**Comment by edenlightning** [comment]
@HareshKarnan, @talolard, @stillwalker1234 or @genghisun can any of you please provide a reproducible script/colab? 

Do you checkpoint based on a value that is not coming from lightning metrics and can be different on different processes? We think there might be an issue there.

**Comment by edenlightning** [comment]
Try setting DDPPlugin(find_unused_parameters=True), seems to fix the issue for other cases.

**Comment by taltalim** [comment]
I'm facing the same issue as described here. Using DDPPlugin(find_unused_parameters=True) does not make a difference, while removing the ModelCheckpoint callback does "solve" the issue. I will try to convert my model to use pl.Metric and see if that helps.

**Comment by JonasFrey96** [comment]
I face the same issue and the problem was resolved by changing the number of workers to 0. This is not an acceptable fix workaround. I assume we have some deadlock when something goes wrong with spaning dataloaders.

**Comment by tchaton** [comment]
Dear @JonasFrey96, @taltalim,@HareshKarnan

Would it be possible for you to work on a reproducible script using the BoringModel.

Best,
T.C

**Comment by jgbos** [comment]
I'm having the same issue that goes away when using the default `ModelCheckpoint` settings.  Seems to be a problem with setting `monitor` and `save_top_k`.  I have not been able to achieve the same issue with `BoringModel`.

**Comment by Anjum48** [comment]
I've been running into this issue with 1.18 & 1.2.1 the last few days with ddp and 2 GPUs

Things tried:
- Upgrade from Pytorch 1.7, to 1.8
- Upgrade drivers from 450 to 460
- Upgrade from CUDA 10.2 to 11.2
- Upgrade NCCL to latest
- Rebuild entire conda env
- Remove all loggers (in case of wandb multithreading)
- os.environ["WANDB_START_METHOD"] = "fork"
- cv2.setNumThreads(0)
- DDPPlugin(find_unused_parameters=True) 
- Reduced num_workers (zero is too slow, but went down to 2)
- Roll back Linux kernel in Ubuntu (I think a saw a meme about this the other day so I thought I'd give it a try haha)
- Nothing unusual in dmesg

The epoch seems to be random (might stop after first, or after several), but always just after validation and on the start of the next epoch. No error messages. Both GPUs locked at "100%" but the data being sent to GPU 0 (RX in nvtop) is 0 MB/s and the temperatures show that the GPUs are not working hard. 2 CPU cores locked at 100%.

**Comment by Anjum48** [comment]
I think I've isolated the issue from the discussion in https://github.com/PyTorchLightning/pytorch-lightning/issues/5604#issuecomment-785783418.

This issue started when I switched the `ModelCheckpoint` monitor from AUC (using PL metrics and `dist_sync_on_step=True`) to `val_loss`. `val_loss` was not being synced between the 2 GPUs, so adding this line:

```
val_loss = torch.mean(self.all_gather(val_loss))
```

In `validation_epoch_end` fixed the issue for me.

**Comment by ifsheldon** [comment]
I can see the same issue when running my training script. My `pl` is 1.2.1. Still trying out which approaches mentioned above is useful in my case.

**Comment by ifsheldon** [comment]
> I think I've isolated the issue from the discussion in [#5604 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/issues/5604#issuecomment-785783418).
> 
> This issue started when I switched the `ModelCheckpoint` monitor from AUC (using PL metrics and `dist_sync_on_step=True`) to `val_loss`. `val_loss` was not being synced between the 2 GPUs, so adding this line:
> 
> ```
> val_loss = torch.mean(self.all_gather(val_loss))
> ```
> 
> In `validation_epoch_end` fixed the issue for me. Model has been successfully training overnight and still running :)

Can you please detail how your `val_loss` is implemented? I think your reasoning is reasonable. But, if I just want some simple metrics(not the pl metric) to be recorded, what can I do?
My validation code is
```python
def validation_epoch_end(self, validation_step_outputs):
# step outputs are a bunch of scalars
        all_outputs = torch.cat(validation_step_outputs)
        std, mean = torch.std_mean(all_outputs)
        self.log("validation_off_mean", mean)
        self.log("validation_off_std", std)
        self.log("validation_off_max", all_outputs.max())
        self.log("validation_off_min", all_outputs.min())
```

**Comment by Anjum48** [comment]
`self.all_gather` will return a tensor of shape `[num_gpus, x]` where x is the result from a single GPU. If you're only syncing a scalar, then you can just take a mean (or min/max/std) like in my example.

If for example, x is not a scalar (e.g. you want to calculate IoU or something), you can combine them using something like:

```
def gather_and_squash(self, x):
        # The reshape goes from [n_gpus, bs, N, N] to [n_gpus * bs, N, N]
        return torch.reshape(self.all_gather(x), [-1] + list(x.shape)[1:])
```

At the moment I'm working on a joint segmentation & classification problem. Here's what my `validation_epoch_end` looks like:
```
def validation_epoch_end(self, outputs):
    loss_val = torch.stack([x["val_loss"] for x in outputs]).mean()
    loss_seg = torch.stack([x["loss_seg"] for x in outputs]).mean()
    loss_clf = torch.stack([x["loss_clf"] for x in outputs]).mean()
    
    log = {
        "loss/valid_seg": torch.mean(self.all_gather(loss_seg)),
        "loss/valid_clf": torch.mean(self.all_gather(loss_clf)),
    }

           
    # AUC calculation etc...


    self.log_dict(log)
    self.log_dict(
        {"loss/valid": loss_val, "auc/overall": log["metric"]}, prog_bar=True,
    )
```

**Comment by jgbos** [comment]
I'm also noticing that there are logger files (tfeevents) for each process now.  I wonder if #6364 and this issue are related.

**Comment by taltalim** [comment]
At the moment I can't seem to reproduce using BoringModel. I will look into that in the next days.

**Comment by senarvi** [comment]
I encountered a similar issue, training hanging at the end of a validation epoch, when a custom metric is being synced between processes. I used `cat` as [dist_reduce_fx](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/metric.py#L89). One would think that in the tensors produced by different processes, the first dimension - along which the tensors are going to be concatenated - doesn't have to match. However, [gather_all_tensors](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/metric.py#L89) uses [torch.distributed.all_gather](https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather) to gather the tensors from different processes, and `all_gather` requires the tensors to be correctly sized. `gather_all_tensors` assumes that all processes create identically-sized tensors. If the sizes don't match, syncing will hang indefinitely.

This is probably not the issue for most people, but I just wanted to point out that this can cause some confusion too. I guess there's no way to check that all processes produce identically-sized tensors. Maybe the Metric API documentation could be updated to note that all state variables must have identical shape across processes.

**Comment by jgbos** [comment]
@Anjum48  did implementing those reduction in `val_epoch_end` fix your issue with training hanging?  My feeling is there's an uncaught error in the syncing of tensors like @senarvi found.   I can't seem to find the error though.

**Comment by jgbos** [comment]
I hope this helps:  I implemented all the reduction and logging in `validation_epoch_end`. 

Looks like PL attempts to do a model checkpoint before running `validation_epoch_end`, the trainer crashes because it was unable to find the monitor metric.  I can verify that `validation_step_end` is called before the crash but not `validation_epoch_end`.   Unfortunately having troubles find the code to check this behavior.

**Comment by TysonYu** [comment]
I have the same problem when using version==1.2.3

**Comment by TysonYu** [comment]
Hi, @ifsheldon  I am facing the same problem, may I ask how do you solve this problem?

**Comment by jgbos** [comment]
An initial test with today's master seems to show this issue is fixed for me

**Comment by thiyagu145** [comment]
looks like this issue is fixed in version 1.2.4

**Comment by jgbos** [comment]
@thiyagu145 oh really? I'll double check, I thought it required some changes not in 1.2.4.  But that would be great if 1.2.4 fixes it.

**Comment by thiyagu145** [comment]
yea, training completed without any issues.

**Comment by TysonYu** [comment]
Yeah, now I try 1.2.4 and there is no issue anymore.

**Comment by sustcsonglin** [comment]
1.2.4 stills

**Comment by taltalim** [comment]
I can confirm 1.2.4 fixes the issue. I'm asking myself which PR fixed this - possibly #6410 ?

**Comment by luozhouyang** [comment]
Same issue.
pytorch-lightning version `1.3.7post0`

**Comment by JusperLee** [comment]
> 同样的问题。
> pytorch-闪电版本`1.3.7post0`

You can downgrade to 1.2.4

**Comment by yinrong** [comment]
I'm stuck too using lightning=1.4.9
![image](https://user-images.githubusercontent.com/4221569/136338235-06d0c777-868c-46a3-9212-acd4fcd3e275.png)

**Comment by HareshKarnan** [comment]
same issue with 1.8.3.post1
